src/cargo/core/resolver/mod.rs

```diff
@@ -247,158 +193,245 @@ pub fn resolve(summary: &Summary, method: &Method,
-           dep.name(), prev_active.len());
-
-    // Filter the set of candidates based on the previously activated
-    // versions for this dependency. We can actually use a version if it
-    // precisely matches an activated version or if it is otherwise
-    // incompatible with all other activated versions. Note that we define
-    // "compatible" here in terms of the semver sense where if the left-most
-    // nonzero digit is the same they're considered compatible.
-    let my_candidates = candidates.iter().filter(|&b| {
-        prev_active.iter().any(|a| a == b) ||
-            prev_active.iter().all(|a| {
-                !compatible(a.version(), b.version())
-            })
-    });
+        let prev_active = cx.prev_active(&dep).to_vec();
+        trace!("{}[{}]>{} {} candidates", parent.name(), cur, dep.name(),
+               candidates.len());
+        trace!("{}[{}]>{} {} prev activations", parent.name(), cur,
+               dep.name(), prev_active.len());
+
+        // Filter the set of candidates based on the previously activated
+        // versions for this dependency. We can actually use a version if it
+        // precisely matches an activated version or if it is otherwise
+        // incompatible with all other activated versions. Note that we define
+        // "compatible" here in terms of the semver sense where if the left-most
+        // nonzero digit is the same they're considered compatible.
+        let my_candidates = candidates.iter().filter(|&b| {
+            prev_active.iter().any(|a| a == b) ||
+                prev_active.iter().all(|a| {
+                    !compatible(a.version(), b.version())
+                })
+        }).cloned().collect();
+
+        // Alright, for each candidate that's gotten this far, it meets the
+        // following requirements:
+        //
+        // 1. The version matches the dependency requirement listed for this
+        //    package
+        // 2. There are no activated versions for this package which are
+        //    semver-compatible, or there's an activated version which is
+        //    precisely equal to `candidate`.
+        //
+        // This means that we're going to attempt to activate each candidate in
+        // turn. We could possibly fail to activate each candidate, so we try
+        // each one in turn.
+        let mut remaining_candidates = RcVecIter::new(my_candidates);
+        let candidate = match remaining_candidates.next() {
+            Some((_, candidate)) => {
+                // We have a candidate. Add an entry to the `backtrack_stack` so
+                // we can try the next one if this one fails.
+                backtrack_stack.push(BacktrackFrame {
+                    context_backup: cx.clone(),
+                    deps_backup: remaining_deps.clone(),
+                    remaining_candidates: remaining_candidates,
+                    parent: parent.clone(),
+                    dep: dep.clone(),
+                });
```

Comment:
Another possible optimization in the future I think is that if `remaining_candidates` is empty then this `push` can almost be elided entirely, the information I believe is only later used to generate an error which may be returned. That being said it's easier to just push for now, so that's fine.